File: src/main/java/gg/climb/lolobjects/esports/Player.java
package gg.climb.lolobjects.esports;

import gg.climb.lolobjects.InternalId;
import gg.climb.lolobjects.RiotId;

import org.immutables.value.Value;

import java.util.Collection;
import java.util.Optional;

/**
 * An actual human player represented in League. Use {@link PlayerBuilder} to get an instance.
 */
@Value.Style(visibility = Value.Style.ImplementationVisibility.PRIVATE,
             strictBuilder = true)
@Value.Immutable
public abstract class Player {
  public abstract InternalId getInternalId();

  public abstract RiotId getRiotId();

  public abstract String getFirstName();

  public abstract String getLastName();

  public abstract String getIgn();

  public abstract Role getRole();

  public abstract String getRegion();

  public abstract Optional<InternalId> getCurrentTeamId();

  public abstract Collection<InternalId> getTeamIds();
}

File: app/gg/climb/lolobjects/esports/Player.scala
package gg.climb.lolobjects.esports

import gg.climb.lolobjects.{InternalId, RiotId}

class Player(val id: InternalId[Player],
             val ign: String,
             val role: Role,
             val riotId: RiotId[Player]) {

  override def toString: String = s"Player(id=$id, ign=$ign, role=$role, riotId=$riotId)"

  override def equals(other: Any): Boolean = other match {
    case that: Player => id == that.id
    case _ => false
  }
  override def hashCode: Int = id.hashCode
}
